Added ability to logically compare LDAPFilterSet objects

mschwager opened the pull request:
Fixes #76 

Here's my purposed fix for the issue I was having in #76. I like it because it's a fairly simple solution to the problem at hand.

It's worth noting, I did consider an alternate solution. Namely, using the `collections.Counter` object. I would've actually preferred this solution because it's faster, but, unfortunately it doesn't work for nested OR and AND filters. E.g. `test_nested_equal`. Something like:

```diff
diff --git a/ldaptor/protocols/pureber.py b/ldaptor/protocols/pureber.py
index 7b12678..9052e13 100644
--- a/ldaptor/protocols/pureber.py
+++ b/ldaptor/protocols/pureber.py
@@ -305,6 +305,9 @@ class BERSequence(BERStructured, UserList.UserList):
             return self.__class__.__name__+"(value=%s, tag=%d)" \
                    %(repr(self.data), self.tag)
 
+    def __hash__(self):
+        return hash(self.asText())
+
 
 class BERSequenceOf(BERSequence):
     pass
diff --git a/ldaptor/protocols/pureldap.py b/ldaptor/protocols/pureldap.py
index f97973c..ef90fcf 100644
--- a/ldaptor/protocols/pureldap.py
+++ b/ldaptor/protocols/pureldap.py
@@ -15,7 +15,7 @@
 
 """LDAP protocol message conversion; no application logic here."""
 
-import operator
+import collections
 import string
 
 from pureber import (
@@ -394,8 +394,10 @@ class LDAPFilterSet(BERSet):
     fromBER = classmethod(fromBER)
 
     def __eq__(self, rhs):
-        return (sorted(self, key=operator.methodcaller('asText')) ==
-            sorted(rhs, key=operator.methodcaller('asText')))
+        return (
+            collections.Counter(self) ==
+            collections.Counter(rhs)
+        )
 
 class LDAPFilter_and(LDAPFilterSet):
     tag = CLASS_CONTEXT|0x00
```

`Counter` uses hashing on the backend, so we'd have to implement `__hash__`, and is preferable because it runs in `O(n)`, but fails at recursive comparisons. `sorted` runs in `O(n log n)`, and works for recursive comparisons because to sort we will recursively be calling `__eq__`.

Note that `Counter` is essentially what [assertItemsEqual](https://docs.python.org/2.7/library/unittest.html#unittest.TestCase.assertItemsEqual) does, but alas won't work recursively :(

Let me know what you think! Especially if you can see some way to make `Counter` work!

psi29a commented:
Thanks for creating this MR, disregard the pypy errors, seems we need to update this.

:1st_place_medal: for the tests :)

mschwager commented:
Ahh, I see. Is there anything I can do to help move forward with this?

cwaldbieser commented:
One simple optimization might be to check if the identity if the left and right hand sides are equal.  E.g. if `id(self) == id(rhs)`, then the 2 items must be equal.  Not sure how much this will help in practice.

cwaldbieser commented:
This is a bit pathological, but by using `asText()` and implicitly using the `self.escaper()`, couldn't 2 filters that are equal end up not comparing the same?  E.g. if I have minimal escaping on one filter but complete filtering on another, their textual representations would be different.

mschwager commented:
Thanks for the feedback.

Per the optimizations, typically I like to keep the code as simple as possible, then tackle optimizations when needed. With that being said, though, doing an identity check first doesn't complicate things too much. Further we could check that their `len` is the same before sorting as well :+1:

As for using `asText`, this point is particularly interesting. At first I was inclined to agree with you that two filters that are otherwise the same, but have differing escape functions, should be considered equal, and the code may not currently represent that. But now I'm leaning the other way. For example, if we have different escape functions we could get the following filter strings from the same object: `(attribute=A)` and `(attribute=01000001)`. Now I believe that these should be considered different. The LDAP directory *should* return the same result if it supports different encodings for the same query, however, say the directory requires the attribute value to be binary encoded. Then the first query would fail and the second would succeed, which fundamentally means the filter objects are different.

Thoughts?

cwaldbieser commented:
Q1: Will the directory service return identical results for 2 filters.
Q2: Are 2 filters semantically identical.

Questions 1 and 2 are different questions.  Sometimes this is subtle, sometimes not.  For example, take  `(name=a*)` and `(name=A*)`.  These filters have a difference in casing, but depending on the schema rules for "name" at the directory, the filters could be equivalent.

I think the best we can hope to do is to answer Q2.  Further, I think we need to assume that [logical equivalences](https://en.wikipedia.org/wiki/Logical_equivalence) are the only differences we could really look at.

Your use case may be even simpler-- I think you just want to see if a filter is equivalent in the case that the terms of a binary AND or OR operation are switched.

I'm not sure if you are interested in something like De Morgan's laws.  E.g. `(!(|(givenName=Carl)(sn=Waldbieser)))` == `(&(!(givenName=Carl))(!(sn=Waldbieser)))` .  Both filters eliminate entries with givenName "Carl" AND entries with sn "Waldbieser".  They are logically equivalent filters, but the forms are rather distinct.

It makes me wonder if overloading the __eq__ operator is the correct kind of comparison.  Maybe a function/method like "test_logical_equivalence()" or a name more appropriate to whatever kind of comparison we are making might be more appropriate.

mschwager commented:
I like your idea of not concerning ourselves with what a directory will consider an equivalent filter. It simplifies much of this, and makes sense since this library shouldn't be trying to anticipate the needs of many different directories.

With that being said, I agree that it would then make sense to avoid the escaping function when considering equivalence. We'll only consider the attribute and value when checking for equivalence and not the escaped/encoded value. I'll make that change.

Further, you are also correct that I'm only trying to find equivalences between simple AND and ORs, not complex logical equivalents, e.g. DeMorgan's. I'll save solving [boolean satisfiability](https://en.wikipedia.org/wiki/Boolean_satisfiability_problem) for another day :)

To your last point, I do particularly like using `__eq__`. It makes comparison Just Work™ with different libraries and operations. For example, [assertEqual](https://docs.python.org/2/library/unittest.html#unittest.TestCase.assertEqual) and other functions that use `==` will just work without having to add special comparators.

I did briefly look into why different escapers were resulting in the same filter. I would've expected otherwise given the predicate `asText`. I did come up with a test to prove your point that there could be incorrect results, I'll add it to the PR. To remedy the issue I've changed to using `str` instead of `asText`. Good catch!

So I've:

* Added some comparison optimizations
* Changed the semantics to not rely on `asText` and thus not rely on the escaper
* Kept `__eq__`, I fairly partial to using it, let me know if this is a dealbreak

Let me know what you think!

cwaldbieser commented:
@psi29a - What do you think?  Operator overloading kind of rubs me the wrong way in this case, but maybe I'm just being too picky?  Otherwise, the PR looks OK to me.

psi29a commented:
Honestly, I think it is a elegant way of handling things, it is self-contained in the overload and there are test-coverage enough that we should be able to track it going forward for regressions.

psi29a commented:
Please add an:
Release 17.1 (UNRELEASED)
section in our NEWS.rst and add your bug-fix.
https://github.com/twisted/ldaptor/blob/master/docs/source/NEWS.rst

After that, we're ready to merge. :)

(Not dumping this on you, I'm asking that any PRs that come in, to add an entry when fixing bugs, adding features or changing the API. That way we don't have to go hunting when making a release.)

mschwager commented:
No sweat, I'm a big fan of maintaining a CHANGELOG. I added this feature to NEWS and also added the functionality I missed from my last PR :)

psi29a commented:
Merging! Thanks again.
